File: 2nd Year/Term 2/Data Structures & Algorithms/MS1/Tree/Pseudocodes/Hash.py
# Hash in Trees - Hash table mapping brand (key) to tree node (MotorPH)


class TreeNode:
    def __init__(self, brand, model, quantity):
        self.data = {"brand": brand, "model": model, "quantity": quantity}
        self.left = None
        self.right = None


def add_to_hash_tree(hash_table, brand, node):
    """Add brand -> node to hash table if brand not already present."""
    if brand not in hash_table:
        hash_table[brand] = node


def get_from_hash_tree(hash_table, brand):
    """Look up node by brand. Returns node or None."""
    return hash_table.get(brand)


def display(hash_table):
    """Display hash table (brand -> node data)."""
    if not hash_table:
        print("  (empty)")
        return
    for brand, node in hash_table.items():
        d = node.data
        print(f"  {brand} -> {d['brand']} {d['model']}: {d['quantity']} units")


def main():
    # Hash table: brand (key) -> TreeNode
    hash_table = {}

    # Create nodes and add to hash tree
    add_to_hash_tree(hash_table, "Honda", TreeNode("Honda", "Click 125", 10))
    add_to_hash_tree(hash_table, "Yamaha", TreeNode("Yamaha", "NMAX 155", 5))
    add_to_hash_tree(hash_table, "Suzuki", TreeNode("Suzuki", "Burgman Street", 7))

    print("Hash table (brand -> node):")
    display(hash_table)

    # Look up by brand
    print("\nGet 'Honda':")
    node = get_from_hash_tree(hash_table, "Honda")
    print(f"  {node.data if node else 'Not found'}")

    print("\nGet 'Kawasaki':")
    node = get_from_hash_tree(hash_table, "Kawasaki")
    print(f"  {node.data if node else 'Not found'}")


if __name__ == "__main__":
    main()

File: 2nd Year/Term 2/Data Structures & Algorithms/MS1/Old files (mistake)/Tree/Pseudocodes/Hash.java
package Tree.Pseudocodes;

import java.util.HashMap;
import java.util.Map;

/**
 * Hash in Trees - Hash table mapping brand (key) to tree node (MotorPH)
 */
public class Hash {

    static class TreeNode {
        InventoryItem data;
        TreeNode left;
        TreeNode right;

        TreeNode(String brand, String model, int quantity) {
            this.data = new InventoryItem(brand, model, quantity);
            this.left = null;
            this.right = null;
        }
    }

    record InventoryItem(String brand, String model, int quantity) {}

    /** Add brand -> node to hash table if brand not already present. */
    static void addToHashTree(Map<String, TreeNode> hashTable, String brand, TreeNode node) {
        if (!hashTable.containsKey(brand)) {
            hashTable.put(brand, node);
        }
    }

    /** Look up node by brand. Returns node or null. */
    static TreeNode getFromHashTree(Map<String, TreeNode> hashTable, String brand) {
        return hashTable.get(brand);
    }

    /** Display hash table (brand -> node data). */
    static void display(Map<String, TreeNode> hashTable) {
        if (hashTable.isEmpty()) {
            System.out.println("  (empty)");
            return;
        }
        for (Map.Entry<String, TreeNode> e : hashTable.entrySet()) {
            InventoryItem d = e.getValue().data;
            System.out.println("  " + e.getKey() + " -> " + d.brand() + " " + d.model() + ": " + d.quantity() + " units");
        }
    }

    public static void main(String[] args) {
        // Hash table: brand (key) -> TreeNode
        Map<String, TreeNode> hashTable = new HashMap<>();

        // Create nodes and add to hash tree
        addToHashTree(hashTable, "Honda", new TreeNode("Honda", "Click 125", 10));
        addToHashTree(hashTable, "Yamaha", new TreeNode("Yamaha", "NMAX 155", 5));
        addToHashTree(hashTable, "Suzuki", new TreeNode("Suzuki", "Burgman Street", 7));

        System.out.println("Hash table (brand -> node):");
        display(hashTable);

        // Look up by brand
        System.out.println("\nGet 'Honda':");
        TreeNode node = getFromHashTree(hashTable, "Honda");
        System.out.println("  " + (node != null ? node.data : "Not found"));

        System.out.println("\nGet 'Kawasaki':");
        node = getFromHashTree(hashTable, "Kawasaki");
        System.out.println("  " + (node != null ? node.data : "Not found"));
    }
}
